File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/VaultLocalKeyProvider.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text.Json;
using HealthFlow.Hfcx.Sdk.Exceptions;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Loads the recipient's RSA private key from a HashiCorp Vault KV v2
/// secret. Sister to Java's <c>VaultLocalKeyProvider</c> and Python's
/// <c>VaultLocalKeyProvider</c>. Cross-SDK invariants:
/// </summary>
/// <remarks>
/// <list type="bullet">
///   <item><c>GET /v1/{mount}/data/{path}</c> with the
///     <c>X-Vault-Token</c> header.</item>
///   <item>Optional <c>X-Vault-Namespace</c> for Vault Enterprise.</item>
///   <item>Configurable <c>secretField</c> defaults to <c>"value"</c>.</item>
///   <item>403 / 404 / missing field → <see cref="KeyUnavailableException"/>.</item>
/// </list>
/// </remarks>
public sealed class VaultLocalKeyProvider : ILocalKeyProvider, IDisposable
{
    /// <summary>Default mount path on the KV-v2 secrets engine.</summary>
    public const string DefaultMount = "secret";

    /// <summary>Default key inside the secret containing the PEM.</summary>
    public const string DefaultSecretField = "value";

    private readonly Uri _vaultBaseUrl;
    private readonly string _secretPath;
    private readonly string _token;
    private readonly string _mount;
    private readonly string _secretField;
    private readonly string? _namespace;
    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;
    private readonly TimeSpan _requestTimeout;

    /// <summary>Construct a Vault-backed key provider.</summary>
    public VaultLocalKeyProvider(
        Uri vaultBaseUrl,
        string secretPath,
        string vaultToken,
        string secretMount = DefaultMount,
        string secretField = DefaultSecretField,
        string? vaultNamespace = null,
        HttpClient? httpClient = null,
        TimeSpan? requestTimeout = null)
    {
        ArgumentNullException.ThrowIfNull(vaultBaseUrl);
        ArgumentException.ThrowIfNullOrEmpty(secretPath);
        ArgumentException.ThrowIfNullOrEmpty(vaultToken);
        ArgumentException.ThrowIfNullOrEmpty(secretMount);
        ArgumentException.ThrowIfNullOrEmpty(secretField);

        _vaultBaseUrl = vaultBaseUrl;
        _secretPath = secretPath;
        _token = vaultToken;
        _mount = secretMount;
        _secretField = secretField;
        _namespace = vaultNamespace;
        _requestTimeout = requestTimeout ?? TimeSpan.FromSeconds(10);

        if (httpClient is null)
        {
            _httpClient = new HttpClient { Timeout = _requestTimeout };
            _ownsHttpClient = true;
        }
        else
        {
            _httpClient = httpClient;
            _ownsHttpClient = false;
        }
    }

    /// <inheritdoc />
    public RSA GetPrivateKey()
    {
        var url = new Uri(_vaultBaseUrl, $"/v1/{_mount}/data/{_secretPath}");
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("X-Vault-Token", _token);
        if (!string.IsNullOrEmpty(_namespace))
        {
            request.Headers.TryAddWithoutValidation("X-Vault-Namespace", _namespace);
        }

        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        HttpResponseMessage? response;
        try
        {
            response = _httpClient.Send(request);
        }
        catch (HttpRequestException ex)
        {
            throw new KeyUnavailableException(
                $"Vault unreachable at '{url}': {ex.Message}");
        }

        try
        {
            if (response.StatusCode is HttpStatusCode.Forbidden or HttpStatusCode.NotFound)
            {
                throw new KeyUnavailableException(
                    $"Vault returned HTTP {(int)response.StatusCode} for '{_secretPath}'");
            }

            if (!response.IsSuccessStatusCode)
            {
                throw new KeyUnavailableException(
                    $"Vault returned HTTP {(int)response.StatusCode} for '{_secretPath}'");
            }

            string body;
            using (var stream = response.Content.ReadAsStream())
            using (var reader = new System.IO.StreamReader(stream))
            {
                body = reader.ReadToEnd();
            }

            var pem = ExtractField(body);

            var rsa = RSA.Create();
            try
            {
                rsa.ImportFromPem(pem);
                return rsa;
            }
            catch (Exception ex) when (ex is ArgumentException or CryptographicException)
            {
                rsa.Dispose();
                throw new KeyUnavailableException(
                    $"Vault secret at '{_secretPath}' did not parse as PKCS#8 PEM: {ex.Message}");
            }
        }
        finally
        {
            response.Dispose();
        }
    }

    private string ExtractField(string body)
    {
        try
        {
            using var doc = JsonDocument.Parse(body);
            if (!doc.RootElement.TryGetProperty("data", out var outerData)
                || !outerData.TryGetProperty("data", out var innerData)
                || !innerData.TryGetProperty(_secretField, out var fieldNode)
                || fieldNode.ValueKind != JsonValueKind.String)
            {
                throw new KeyUnavailableException(
                    $"Vault secret at '{_secretPath}' is missing required field '{_secretField}'");
            }

            return fieldNode.GetString()!;
        }
        catch (JsonException ex)
        {
            throw new KeyUnavailableException(
                $"Vault response for '{_secretPath}' was not valid JSON: {ex.Message}");
        }
    }

    /// <inheritdoc />
    public void Dispose()
    {
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
    }
}

File: sdk-javascript/src/recipient/VaultLocalKeyProvider.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { type KeyLike, importPKCS8 } from 'jose';

import type { FetchFn } from '../auth/KeycloakTokenClient.js';
import { JWE_ALG } from '../crypto/JweAlgorithms.js';
import { KeyUnavailableError } from '../exceptions/HfcxError.js';
import type { LocalKeyProvider } from './LocalKeyProvider.js';

export interface VaultLocalKeyProviderOptions {
  vaultBaseUrl: string | URL;
  secretPath: string;
  vaultToken: string;
  /** Default `'secret'`. */
  secretMount?: string;
  /** Default `'value'`. */
  secretField?: string;
  /** Optional Vault namespace. */
  vaultNamespace?: string;
  /** Per-request timeout. Default 10 s. */
  requestTimeoutMs?: number;
  /** Test seam: inject a fetch (defaults to `globalThis.fetch`). */
  fetchFn?: FetchFn;
}

/**
 * Loads the recipient's RSA private key from a HashiCorp Vault KV v2
 * secret. Sister to Java's `VaultLocalKeyProvider`, Python's
 * `VaultLocalKeyProvider`, and .NET's `VaultLocalKeyProvider`.
 *
 *   - GET `/v1/{mount}/data/{path}` with the `X-Vault-Token` header.
 *   - Optional `X-Vault-Namespace` for Vault Enterprise.
 *   - Configurable `secretField`, defaulting to `"value"`.
 *   - 403 / 404 / missing field → {@link KeyUnavailableError}.
 */
export class VaultLocalKeyProvider implements LocalKeyProvider {
  static readonly DEFAULT_MOUNT = 'secret';
  static readonly DEFAULT_SECRET_FIELD = 'value';

  private readonly url: URL;
  private readonly secretPath: string;
  private readonly token: string;
  private readonly secretField: string;
  private readonly namespace: string | undefined;
  private readonly fetchFn: FetchFn;
  private readonly requestTimeoutMs: number;

  constructor(options: VaultLocalKeyProviderOptions) {
    if (!options.vaultBaseUrl) throw new TypeError('vaultBaseUrl is required');
    if (!options.secretPath) throw new TypeError('secretPath is required');
    if (!options.vaultToken) throw new TypeError('vaultToken is required');

    const base =
      options.vaultBaseUrl instanceof URL ? options.vaultBaseUrl : new URL(options.vaultBaseUrl);
    const mount = options.secretMount ?? VaultLocalKeyProvider.DEFAULT_MOUNT;
    this.url = new URL(`/v1/${mount}/data/${options.secretPath}`, base);
    this.secretPath = options.secretPath;
    this.token = options.vaultToken;
    this.secretField = options.secretField ?? VaultLocalKeyProvider.DEFAULT_SECRET_FIELD;
    this.namespace = options.vaultNamespace;
    this.fetchFn = options.fetchFn ?? globalThis.fetch;
    this.requestTimeoutMs = options.requestTimeoutMs ?? 10_000;
  }

  async getPrivateKey(): Promise<KeyLike> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.requestTimeoutMs);
    try {
      const headers: Record<string, string> = {
        'X-Vault-Token': this.token,
        Accept: 'application/json',
      };
      if (this.namespace) headers['X-Vault-Namespace'] = this.namespace;

      let response: Response;
      try {
        response = await this.fetchFn(this.url, {
          method: 'GET',
          headers,
          signal: controller.signal,
        });
      } catch (err) {
        throw new KeyUnavailableError(
          `Vault unreachable at '${this.url}': ${(err as Error).message ?? String(err)}`,
        );
      }

      if (response.status === 403 || response.status === 404 || !response.ok) {
        throw new KeyUnavailableError(
          `Vault returned HTTP ${response.status} for '${this.secretPath}'`,
        );
      }

      let body: unknown;
      try {
        body = await response.json();
      } catch (err) {
        throw new KeyUnavailableError(
          `Vault response for '${this.secretPath}' was not valid JSON: ${(err as Error).message}`,
        );
      }

      const pem = extractField(body, this.secretField);
      if (!pem) {
        throw new KeyUnavailableError(
          `Vault secret at '${this.secretPath}' is missing required field '${this.secretField}'`,
        );
      }

      try {
        return await importPKCS8(pem, JWE_ALG);
      } catch (err) {
        throw new KeyUnavailableError(
          `Vault secret at '${this.secretPath}' did not parse as PKCS#8 PEM: ${
            (err as Error).message
          }`,
        );
      }
    } finally {
      clearTimeout(timer);
    }
  }
}

function extractField(body: unknown, field: string): string | undefined {
  if (!body || typeof body !== 'object') return undefined;
  const outer = (body as { data?: unknown }).data;
  if (!outer || typeof outer !== 'object') return undefined;
  const inner = (outer as { data?: unknown }).data;
  if (!inner || typeof inner !== 'object') return undefined;
  const value = (inner as Record<string, unknown>)[field];
  return typeof value === 'string' ? value : undefined;
}
